### ISSUE TITLE
ConcurrentModificationException in TwoLevelTileCache.setWorkingSet

### ISSUE THREAD

**Issue by vennekamp** [mapsforge/mapsforge#649]
I'm experiencing an exception on a map with many additonal layers (markers):
java.util.ConcurrentModificationException
at java.util.HashMap$HashIterator.nextEntry(HashMap.java:806)
at java.util.HashMap$KeyIterator.next(HashMap.java:833)
at 
org.mapsforge.map.layer.cache.TwoLevelTileCache.setWorkingSet(TwoLevelTileCache.java:101)
at org.mapsforge.map.layer.TileLayer.draw(TileLayer.java:87)
at org.mapsforge.map.layer.LayerManager.doWork(LayerManager.java:90)
at org.mapsforge.map.util.PausableThread.run(PausableThread.java:140) 
 
see also our discussion in google groups:
https://groups.google.com/forum/#!topic/mapsforge-dev/xBNatQkhkCI 

I have the following settings in mappreferences.xml: android:key="tilecache_threading";  android:defaultValue="true" />

**Comment by devemux86** [comment]
@vennekamp in the forum you mentioned you used old Mapsforge with the deprecated threaded tile cache.

Did you see any relevant problems with latest Mapsforge and regular cache?

**Comment by vennekamp** [comment]
I have to check, will takesome time (I did not yet upgrade to 0.6) - Best regards, Martin

**Comment by mvglasow** [comment]
Same here with official 0.6.0 binaries:

```
04-04 18:28:53.009: E/AndroidRuntime(6483): FATAL EXCEPTION: LayerManager
04-04 18:28:53.009: E/AndroidRuntime(6483): Process: com.vonglasow.michael.satstat, PID: 6483
04-04 18:28:53.009: E/AndroidRuntime(6483): java.util.ConcurrentModificationException
04-04 18:28:53.009: E/AndroidRuntime(6483): 	at java.util.HashMap$HashIterator.nextEntry(HashMap.java:787)
04-04 18:28:53.009: E/AndroidRuntime(6483): 	at java.util.HashMap$KeyIterator.next(HashMap.java:814)
04-04 18:28:53.009: E/AndroidRuntime(6483): 	at org.mapsforge.map.layer.cache.TwoLevelTileCache.setWorkingSet(TwoLevelTileCache.java:98)
04-04 18:28:53.009: E/AndroidRuntime(6483): 	at org.mapsforge.map.layer.TileLayer.draw(TileLayer.java:88)
04-04 18:28:53.009: E/AndroidRuntime(6483): 	at org.mapsforge.map.layer.download.TileDownloadLayer.draw(TileDownloadLayer.java:57)
04-04 18:28:53.009: E/AndroidRuntime(6483): 	at org.mapsforge.map.layer.LayerManager.doWork(LayerManager.java:90)
04-04 18:28:53.009: E/AndroidRuntime(6483): 	at org.mapsforge.map.util.PausableThread.run(PausableThread.java:140)
```

Code to create the tile cache:
```
mapTileCache = AndroidUtil.createExternalStorageTileCache(rootView.getContext(),
		"MapQuest",
		Math.round(AndroidUtil.getMinimumCacheSize(rootView.getContext(),
				mapMap.getModel().displayModel.getTileSize(),
				mapMap.getModel().frameBufferModel.getOverdrawFactor(),
				1f)),
		mapMap.getModel().displayModel.getTileSize(),
		true);
```

Additional markers are present. The error occurred as I dragged the map and then jumped back to the previous position programatically. It occurs only sporadically, as is the case with concurrent modifications and other race conditions.

Admittedly, I haven't been following progress on MapsForge for over a year and the API in question has changed quite a bit since – should the above call create a "regular" (non-threaded) cache?

**Comment by devemux86** [comment]
Threaded tile cache was removed before that.

The code you mention creates a persistent tile cache.
You can test `dev` code and see if having the same behavior.

**Comment by mvglasow** [comment]
Just to clarify: in dev@454c97b, both flavors of `AndroidUtil.createExternalStorageTileCache()` appear to take two extra arguments, `threaded` and `queueSize`, which are not present in 0.6.0 (they seem to have been dropped from 0.5.1 to 0.5.2 as per the API docs). With the call I posted earlier, Eclipse will complain that there is no method with a matching signature. However, the API docs for dev make no mention of these extra arguments.

Were these arguments (re-)introduced in `dev`, or should they have been removed but were forgotten, are the dev API docs out of date, or have I inadvertently built from old sources?
Is that intentional, or is there a mistake somewhere?

**Comment by mvglasow** [comment]
dev@454c97b, same thing here:

```
04-04 22:21:43.346: E/AndroidRuntime(24691): FATAL EXCEPTION: LayerManager
04-04 22:21:43.346: E/AndroidRuntime(24691): Process: com.vonglasow.michael.satstat, PID: 24691
04-04 22:21:43.346: E/AndroidRuntime(24691): java.util.ConcurrentModificationException
04-04 22:21:43.346: E/AndroidRuntime(24691): 	at java.util.HashMap$HashIterator.nextEntry(HashMap.java:787)
04-04 22:21:43.346: E/AndroidRuntime(24691): 	at java.util.HashMap$KeyIterator.next(HashMap.java:814)
04-04 22:21:43.346: E/AndroidRuntime(24691): 	at org.mapsforge.map.layer.cache.TwoLevelTileCache.setWorkingSet(TwoLevelTileCache.java:98)
04-04 22:21:43.346: E/AndroidRuntime(24691): 	at org.mapsforge.map.layer.TileLayer.draw(TileLayer.java:88)
04-04 22:21:43.346: E/AndroidRuntime(24691): 	at org.mapsforge.map.layer.download.TileDownloadLayer.draw(TileDownloadLayer.java:57)
04-04 22:21:43.346: E/AndroidRuntime(24691): 	at org.mapsforge.map.layer.LayerManager.doWork(LayerManager.java:94)
04-04 22:21:43.346: E/AndroidRuntime(24691): 	at org.mapsforge.map.util.PausableThread.run(PausableThread.java:140)
```

**Comment by mvglasow** [comment]
For anyone wishing to reproduce or test a potential fix: 

* clone https://github.com/mvglasow/satstat
* `git revert 9796b0a`
* replace mapsforge jars in `libs` with those from latest dev snapshot
* import code into new Eclipse project using ADT or Andmore (don't get distracted by `build.gradle`, I maintain it for F-Droid but Eclipse doesn't use it)
* launch SatStat
* go to the map tab and pan around for a while (may take a few minutes)
* if SatStat crashes, it will write a crash log similar to the one above to `Android/data/com.vonglasow.michael.satstat/files` (you need to do a media scan before the latest log is visible over MTP).

**Comment by mvglasow** [comment]
Taking a quick look on the code, the offender seems to be

    org.mapsforge.map.layer.cache.TwoLevelTileCache.setWorkingSet(TwoLevelTileCache.java:98)
        @Override
        public void setWorkingSet(Set<Job> newWorkingSet) {
            this.workingSet.clear();
            this.workingSet.addAll(newWorkingSet);
            this.firstLevelTileCache.setWorkingSet(this.workingSet);
            this.secondLevelTileCache.setWorkingSet(this.workingSet);
    >       for (Job job : workingSet) {

Apparently accesses to `workingSet` are not synchronized properly. Further up in the same file, in the constructor, I find

    this.workingSet = Collections.synchronizedSet(new HashSet<Job>());

which, according to the docs, wraps a `HashSet<Job>` in a thread-safe way and thus should prevent the very kind of error we're seeing here. However, taking a look at the next item in the call stack, I find

    org.mapsforge.map.layer.TileLayer.draw(TileLayer.java:88)
        Set<Job> jobs = new HashSet<>();
        for (TilePosition tilePosition : tilePositions) {
            jobs.add(createJob(tilePosition.tile));
        }
    >    this.tileCache.setWorkingSet(jobs);

So here we're creating a fresh (non-synchronized) `HashSet`, populating it with jobs and passing it to `setWorkingSet()`. Looking at that function again, we see that it clears the existing `workingSet`, then adds the new set.

Is there any way a synchronized set can lose its thread-safe properties (e.g. by clearing its contents and adding a non-thread-safe set to it, as we do in the code snippet above)?

**Comment by mvglasow** [comment]
I've found something interesting at http://docs.oracle.com/javase/6/docs/api/java/util/Collections.html#synchronizedSet%28java.util.Set%29:

> ### synchronizedSet
> [...]
>  It is imperative that the user manually synchronize on the returned set when iterating over it:
```
  Set s = Collections.synchronizedSet(new HashSet());
      ...
  synchronized(s) {
      Iterator i = s.iterator(); // Must be in the synchronized block
      while (i.hasNext())
          foo(i.next());
  }
> ``` 
> Failure to follow this advice may result in non-deterministic behavior.

Looks like that is the issue here.

**Comment by devemux86** [comment]
Thanks for the info.
Since you can reproduce the problem, can you try the `synchronized` and see if it solves the issue?

**Comment by mvglasow** [comment]
Sure, I'll report back as soon as I know more.

**Comment by mvglasow** [comment]
Well... as is the case with sporadic bugs, I didn't manage to reproduce it any more using current dev and the latest build of SatStat, so I can't tell for sure if this actually fixes the bug.

In any case, I understand that this iteration is not thread-safe unless explicitly synchronized. I've therefore made the proposed modifications and built SatStat against the updated libraries. Things look OK so far – that is, no obvious regressions.

I'll keep testing it in the next days to verify the issue doesn't recur. In the meantime, I'll file a merge request.

**Comment by devemux86** [comment]
I merged the PR, let's keep an eye on it (also on rendering speed).

@vennekamp if you can test `dev` too.

**Comment by mvglasow** [comment]
Just had another crash, but in a different portion of the code:

```
java.util.ConcurrentModificationException
	at java.util.HashMap$HashIterator.nextEntry(HashMap.java:787)
	at java.util.HashMap$KeyIterator.next(HashMap.java:814)
	at org.mapsforge.core.util.WorkingSetCache.setWorkingSet(WorkingSetCache.java:41)
	at org.mapsforge.map.layer.cache.InMemoryTileCache.setWorkingSet(InMemoryTileCache.java:123)
	at org.mapsforge.map.layer.cache.TwoLevelTileCache.setWorkingSet(TwoLevelTileCache.java:96)
	at org.mapsforge.map.layer.TileLayer.draw(TileLayer.java:88)
	at org.mapsforge.map.layer.download.TileDownloadLayer.draw(TileDownloadLayer.java:57)
	at org.mapsforge.map.layer.LayerManager.doWork(LayerManager.java:94)
	at org.mapsforge.map.util.PausableThread.run(PausableThread.java:140)
```

Apparently `WorkingSetCache.setWorkingSet()` is suffering from the same issue. Will investigate...

**Comment by devemux86** [comment]
Probably `FileSystemTileCache.setWorkingSet`, which has also iteration, should be handled the same way.

**Comment by mvglasow** [comment]
Indeed, I went through all classes implementing `TileCache` and checked for unsynchronized iterations. `WorkingSetCache` and `FileSystemTileCache` were the only ones I found; the referenced merge request fixes them. Should such iterations be present in other parts of the code, they would probably need fixing as well.

**Comment by jvoyatz** [comment]
Is this bug solved? Where can i find the jars or i have to build them on my own?

**Comment by devemux86** [comment]
We published Mapsforge 0.6.1-rc1 containing the fix.
You can use it via [Gradle / Maven](https://github.com/mapsforge/mapsforge/blob/master/docs/Integration.md) or [download](https://github.com/mapsforge/mapsforge/blob/master/docs/Downloads.md) jars.
